### PR TITLE
lib/attrsets: add mapCartesianProduct function

### PR DESCRIPTION
**Pull request by gvolpe**
## Description of changes

The `cartesianProductOfSets` function was [added some time ago](https://github.com/NixOS/nixpkgs/pull/110787) (aka ~`transpose` or~ `sequence` in other functional languages), while deprecating the `crossLists` function. I have come across the deprecation message "use cartesianProductOfSets instead" without further details, which motivated me to improve the situation by adding a new convenient function. Example:

```nix
λ lib.mapCartesianProduct ({a,b}: a+b) {a=[1 2]; b=[3 4];}
[ 4 5 5 6 ]
```

It's the equivalent of using the former followed by `map`, which seems to be quite a common pattern to justify a single function:

```nix
λ builtins.map ({a,b}: a+b) (lib.cartesianProductOfSets {a=[1 2]; b=[3 4];})
[ 4 5 5 6 ]
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by gvolpe** [comment]
Naming is hard, so happy to accept better suggestions! 😄 

I'm now thinking `transposeWith` might be a better name, but keen to hear what others think first.

**Comment by infinisil** [comment]
> aka `transpose` or `sequence` in other functional languages

I don't think I've ever seen this operation be called that? Transposing usually means diagonally mirroring a Matrix.

`cartesianProductOfSets` is a bit too long imo, I think we could shorten it to `cartesianProduct` though, otherwise it seems like a fitting name.

And then I thin `mapCartesianProduct` would be best because it's the same as `map f . cartesianProduct a`.

**Comment by gvolpe** [comment]
> I don't think I've ever seen this operation be called that? Transposing usually means diagonally mirroring a Matrix.

Oh I got the `transpose` function completely wrong, you're right it doesn't apply to this use case 😅 

Though, `sequence` is what it's called in functional languages like Haskell and Scala, e.g.

```haskell
λ sequence [[1,2,3],[4,5,6]]
[[1,4],[1,5],[1,6],[2,4],[2,5],[2,6],[3,4],[3,5],[3,6]]
```

It is [more](https://hackage.haskell.org/package/base-4.19.1.0/docs/Prelude.html#v:sequence) [generic](https://github.com/typelevel/cats/blob/3d3f63abe4fcf7d778c9eb3e695f361f72d24c2e/core/src/main/scala/cats/Traverse.scala#L106) than this, but it results in the Cartesian product of every item when dealing with lists of lists.

> cartesianProductOfSets is a bit too long imo, I think we could shorten it to cartesianProduct though, otherwise it seems like a fitting name.

I agree. This would mean deprecating the existing `cartesianProductOfSets` for this new one, right?

> And then I thin mapCartesianProduct would be best because it's the same as map f . cartesianProduct a.

Definitely makes sense, would be happy to adjust to the suggested names, thanks for reviewing!

**Comment by infinisil** [comment]
> Though, `sequence` is what it's called in functional languages like [Haskell](https://hackage.haskell.org/package/base-4.19.1.0/docs/Prelude.html#v:sequence) and [Scala](https://github.com/typelevel/cats/blob/3d3f63abe4fcf7d778c9eb3e695f361f72d24c2e/core/src/main/scala/cats/Traverse.scala#L106), e.g.
> 
> ```haskell
> λ sequence [[1,2,3],[4,5,6]]
> [[1,4],[1,5],[1,6],[2,4],[2,5],[2,6],[3,4],[3,5],[3,6]]
> ```
> 
> It is more generic than this, but it results in the Cartesian product of every item when dealing with lists of lists.

I'd say that's half of a coincidence, because it relies on the `Applicative` instance (see the more generic [`sequenceA`](https://hackage.haskell.org/package/base-4.19.1.0/docs/Data-Traversable.html#v:sequenceA)) for lists to be the cartesian product, but there's also [another `Applicative` implementation](https://en.wikibooks.org/wiki/Haskell/Applicative_functors#ZipList) on lists that would make sense :D

> I agree. This would mean deprecating the existing `cartesianProductOfSets` for this new one, right?

Yeah I guess that would be best.

File: lib/attrsets.nix [hunk 1]
```diff
@@ -1999,4 +2017,8 @@ rec {
   # DEPRECATED
   zip = warn
     "lib.zip is a deprecated alias of lib.zipAttrsWith." zipAttrsWith;
+
+  # DEPRECATED
+  cartesianProductOfSets = warn
```
Comment (infinisil) [inline lib/attrsets.nix:2022]:
```suggestion
  cartesianProductOfSets = warnIf (isInOldestRelease 2405)
```

See https://github.com/NixOS/nixpkgs/issues/291939 why that's necessary

File: lib/lists.nix [hunk 2]
```diff
@@ -1688,17 +1688,22 @@ rec {
...
+
+    ```nix
+    mapCartesianProduct (x: "${toString x.a}${toString x.b}") { a = [1 2]; b = [3 4]; }
+    => [ "13" "14" "23" "24" ]
+    ```
     :::
   */
   crossLists = warn
-    "lib.crossLists is deprecated, use lib.cartesianProductOfSets instead."
+    "lib.crossLists is deprecated, use lib.mapCartesianProduct instead."
```
Comment (infinisil) [inline lib/lists.nix:1704]: I guess if we want to actually make sure that people understand how to use a new function, this should come with an example. Though in this case, `crossLists` is already deprecated for a while.

**Review by infinisil — CHANGES_REQUESTED** [review]
Please also separate each change into a separate commit, e.g.:
- Renaming `cartesianProductOfSets`
- Tree-wide replacement
- Introducing `mapCartesianProduct` and introducing its tests
- Tree-wide replacement
- Applying lints

File: lib/attrsets.nix [hunk 3]
```diff
@@ -913,6 +913,24 @@ rec {
     ) [{}] (attrNames attrsOfLists);
 
 
+  /**
+    Return the result of function f applied to the cartesian product of attribute set value combinations.
+    Equivalent to using cartesianProduct followed by map.
+
```
Comment (infinisil) [inline lib/attrsets.nix:919]: Needs an `Inputs` and `Type` section, see the other functions in this file (though this is [very recent](https://github.com/NixOS/nixpkgs/pull/296186))

**Comment by gvolpe** [comment]
Should've looked at your profile before, we could discuss ZipLists another day 😄 

> Applying lints

Can you elaborate on what this means? I guess it's running some kind of formatter, but I'm not familiar with the current process.

File: lib/lists.nix (hunk 2 above)
Comment (gvolpe) [reply lib/lists.nix:1704]: The updated documentation shows how to migrate from `crossLists` to `mapCartesianProduct`, is it usually in both places (the warn message and the docs)? Happy to add it, it's been a while since I've contributed to nixpkgs :D

Comment (gvolpe) [reply lib/lists.nix:1704]: If that's the case, do we use something like this for multi-line warning messages (using double `'` single quotes)?

```nix
 crossLists = warn ''
  lib.crossLists is deprecated, use lib.mapCartesianProduct instead.

  To migrate to `lib.mapCartesianProduct` we need to pass an attrset as input instead of a list:

  ```nix
  mapCartesianProduct (x: "${toString x.a}${toString x.b}") { a = [1 2]; b = [3 4]; }
  => [ "13" "14" "23" "24" ]
  \```
''
```

File: lib/tests/misc.nix [hunk 4]
```diff
@@ -1379,7 +1378,7 @@ runTests {
         there
         test'';
     };
-    expected = rec {
+    expected = {
```
Comment (infinisil) [inline lib/tests/misc.nix:1381]: Changes like these should be at least in a separate commit

File: lib/lists.nix (hunk 2 above)
Comment (gvolpe) [reply lib/lists.nix:1704]: @infinisil sounds good, but where? Both in the documentation and the warn message? One or the other?

Comment (gvolpe) [reply lib/lists.nix]: Added an example in both places. Here's how it looks in the REPL:

crossLists-deprecation

Let me know if you're happy with it, I'll proceed with getting the commits right as suggested.

**Comment by gvolpe** [comment]
@infinisil could you have a look at the remaining open questions when you get a chance? Thanks

File: lib/lists.nix (hunk 2 above)
Comment (infinisil) [reply lib/lists.nix]: Sounds good to me!

**Review by infinisil — APPROVED** [review]
Other than the [commits needing cleanup](https://github.com/NixOS/nixpkgs/pull/298680#pullrequestreview-1958799006) this looks good to me!

**Comment by gvolpe** [comment]
>Other than the commits needing cleanup this looks good to me!

@infinisil done ✅

**Comment by Mindavi** [comment]
LGTM except for the 'apply lints' commit that doesn't seem very related

File: lib/tests/misc.nix [hunk 5]
```diff
@@ -1468,13 +1466,10 @@ runTests {
     expected  = "«foo»";
   };
 
-  testToPlist =
-    let
-      deriv = derivation { name = "test"; builder = "/bin/sh"; system = "aarch64-linux"; };
```
[the remark this answers is not in the document]
Comment (gvolpe) [reply lib/tests/misc.nix:1472]: @Mindavi that was suggested [here](https://github.com/NixOS/nixpkgs/pull/298680#pullrequestreview-1958799006). It's removing an unused binding in this case, which come from the linter.

**Comment by gvolpe** [comment]
> LGTM except for the 'apply lints' commit that doesn't seem very related

It is not related indeed, but I had to make changes on that file and noticed a lot of things that can be removed. I don't think anybody would care to submit a PR to remove unused bindings to be honest, but happy to adjust the commit message if there are better suggestions.

**Comment by Mindavi** [comment]
I think it's ok on second thought, thanks!
